File: src/main/java/net/lomeli/trophyslots/core/command/CommandTrophySlots.java
package net.lomeli.trophyslots.core.command;

import java.util.ArrayList;
import java.util.List;

import net.minecraft.command.CommandBase;
import net.minecraft.command.ICommandSender;
import net.minecraft.util.ChatComponentTranslation;

public class CommandTrophySlots extends CommandBase {
    private static List<CommandBase> modCommands = new ArrayList<CommandBase>();
    private static List<String> commands = new ArrayList<String>();

    static {
        modCommands.add(new CommandSetSlots());
        modCommands.add(new CommandGetSlots());
        modCommands.add(new CommandUnlockAll());
        modCommands.add(new CommandUnlockSlots());
        modCommands.add(new CommandRemoveAll());
        modCommands.add(new CommandRemoveSlots());

        for (CommandBase commandBase : modCommands) {
            commands.add(commandBase.getCommandName());
        }
    }

    @Override
    public String getCommandName() {
        return "tslots";
    }

    @Override
    public int getRequiredPermissionLevel() {
        return 0;
    }

    @Override
    public String getCommandUsage(ICommandSender sender) {
        return "command.trophyslots.usage";
    }

    @Override
    public void processCommand(ICommandSender sender, String[] args) {
        if (args.length >= 1) {
            for (CommandBase command : modCommands) {
                if (command.getCommandName().equalsIgnoreCase(args[0]) && command.canCommandSenderUseCommand(sender))
                    command.processCommand(sender, args);
            }
        } else
            sender.addChatMessage(new ChatComponentTranslation("command.trophyslots.usage"));
    }

    @Override
    public List addTabCompletionOptions(ICommandSender commandSender, String[] args) {
        if (args.length == 1)
            return getListOfStringsFromIterableMatchingLastWord(args, commands);
        else if (args.length >= 2) {
            for (CommandBase command : modCommands) {
                if (command.getCommandName().equalsIgnoreCase(args[0]))
                    return command.addTabCompletionOptions(commandSender, args);
            }
        }

        return null;
    }
}

File: src/main/kotlin/net/lomeli/trophyslots/core/command/CommandTrophySlots.kt
package net.lomeli.trophyslots.core.command

import com.google.common.collect.Lists
import net.minecraft.command.CommandBase
import net.minecraft.command.ICommandSender
import net.minecraft.server.MinecraftServer
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentTranslation

class CommandTrophySlots : CommandBase {
    private val modCommands = Lists.newArrayList<CommandBase>()
    private val commands = Lists.newArrayList<String>()

    constructor() {
        modCommands.add(CommandUnlockAll())
        modCommands.add(CommandGetSlots())
        modCommands.add(CommandRemoveAll())
        modCommands.add(CommandRemoveSlots())
        modCommands.add(CommandSetSlots())
        modCommands.add(CommandUnlockSlots())

        var i = 0;
        while (i < modCommands.size) {
            commands.add(modCommands[i].name)
            ++i
        }
    }

    override fun getName(): String? = "tslots"

    override fun getRequiredPermissionLevel(): Int = 0

    override fun getUsage(sender: ICommandSender?): String? = "command.trophyslots.usage"

    override fun execute(server: MinecraftServer?, sender: ICommandSender?, args: Array<out String>?) {
        if (sender != null) {
            if (args != null && args.size >= 1) {
                for (i in modCommands.indices) {
                    val commandBase: CommandBase = modCommands[i];
                    if (commandBase.name.equals(args[0], true) && commandBase.checkPermission(server, sender))
                        commandBase.execute(server, sender, args)
                }
            } else
                sender.sendMessage(TextComponentTranslation("command.trophyslots.usage"))
        }
    }

    override fun getTabCompletions(server: MinecraftServer?, sender: ICommandSender?, args: Array<out String>?, pos: BlockPos?): MutableList<String> {
        if (sender != null && args != null) {
            if (args.size == 1)
                return CommandBase.getListOfStringsMatchingLastWord(args, commands)
            else if (args.size >= 2) {
                for (i in modCommands.indices) {
                    val command: CommandBase = modCommands[i];
                    if (command.name.equals(args[0], true))
                        return command.getTabCompletions(server, sender, args, pos)
                }
            }
        }
        return super.getTabCompletions(server, sender, args, pos)
    }
}